### ISSUE TITLE
WiFi connection errors on startup

### ISSUE THREAD

**Issue by frankonski** [HomeSpan/HomeSpan#896]
I have been wondering why, after startup, I kept getting these errors logged in the Serial console:
```
Trying to connect to My Wifi Network.  Waiting 1 sec...
Trying to connect to My Wifi Network.  Waiting 2 sec...
[  4756][E][WiFiGeneric.cpp:113] set_esp_interface_ip(): Netif Set IP Failed! 0x5007
Trying to connect to My Wifi Network.  Waiting 4 sec...
```
then up to more than 16 seconds, then a minute... btw, `0x5007` is `ESP_ERR_ESP_NETIF_DHCP_NOT_STOPPED`
.. or this one:
```
Trying to connect to My Wifi Network.  Waiting 2 sec...
E (11070) wifi:sta is connecting, return error
[  5381][E][WiFiSTA.cpp:317] begin(): connect failed! 0x3007
Trying to connect to My Wifi Network.  Waiting 4 sec...
E (12459) wifi:sta is connecting, return error
[  6770][E][WiFiSTA.cpp:317] begin(): connect failed! 0x3007
WiFi Connected!  IP Address = 192.168.80.118
```
.. After logging the return value of `WiFi.begin()` in `HomeSpan.cpp`, we see it immediately return `WL_DISCONNECTED`, which is a status set by the WiFi arduino library _during_ the connection attempt.

The issue seems to be that HS is currently calling `WiFi.begin()` back-to-back, too fast, causing some strange behavior. 
Adding `WiFi.waitForConnectResult(20000)` right after solves this issue:
```
    if(waitTime==32000){
      LOG0("\n*** Can't connect to %s.  You may type 'W <return>' to re-configure WiFi, or 'X <return>' to erase WiFi credentials.  Will try connecting again in 60 seconds.\n\n",network.wifiData.ssid);
      waitTime=60000;
    } else {    
      if(verboseWifiReconnect)
        addWebLog(true,"Trying to connect to %s.  Waiting %d sec...",network.wifiData.ssid,waitTime/1000);
      WiFi.begin(network.wifiData.ssid,network.wifiData.pwd);
      // Wait 20 seconds maximum for a successful or failed connection
      WiFi.waitForConnectResult(20000);
    }
```
Of course, right now 20 seconds is hardcoded, but this is just a connect-timeout value which ideally would be configurable. On my network, this method will exit between 3 and 5.5 seconds, and I no longer see any error on connections.

Any chance to add this fix?

**Comment by HomeSpan** [comment]
This whole section is going to be redone when I add support for Ethernet, so I'd like to keep all the code fixed in current state for now.

**Comment by frankonski** [comment]
There I was hoping to get one more contribution to this wonderful project.. ;-)

Feel free to close then, I will keep this fix in my fork, as it happened to me that my device hits that 16 seconds mark and won't connect for a minute. With this one-liner, wifi connection is guaranteed to be fast for me.

**Comment by frankonski** [comment]
Is the fix in the current 1.10.0 branch? I'm seeing the same issue.

**Comment by HomeSpan** [comment]
No, as noted previously, all of this code will be rewritten in a future release. Apart from changes to Pixel that were required to get things to run under IDF 5, I plan on releasing 1.10.0 with no other changes to any other part of the code.  I may rename this 2.0.0 instead since even though it is backwards-compatible with existing HomeSpan, it takes up so much space that it requires re-partitioning which can't be done OTA (only serial). 

Once that is done I will be able to turn attention to improvements, primarily focused on adding Ethernet and revising WiFi (which will address your issue) and modularizing things in a way that will hopefully reduce the footprint.  

Until then, I'm not going to able to add anything else, even small stuff.

**Comment by frankonski** [comment]
well, ultimately you are the owner of this repo. Effort vs impact, this one-liner above to prevent more issues to be created ("_HomeSpan becomes unresponsive after connecting to network_") may have kept the _Issues_ tab a bit quieter.
I'll keep the fix in my fork for now then.

.. and good call about the 2.0. One thing that could be added is in weblogs (or serial) at startup, when you print the HS status, you could potentially print a warning if the size of the "next OTA" partition is smaller than "current size + expected growth", where expected growth is the growth you have seen by moving to arduino-esp32 3.x.

**Comment by HomeSpan** [comment]
I've started work on version 2.1.0 and the first task was to refactor the WiFi connectivity code as noted above.  New code is now available in the dev branch and includes a method to configure the wait times when connecting.  This should solve your issue above and allow you to start with an initial longer time period.   Note that using WiFi.waitForConnectResult() is not desirable since it blocks.  A lot of the logic in HomeSpan is designed to allow it to start operating WHILE it is trying to connect.  This ensures immediate and continuous manual operation of lights/switches when a device is rebooted or WiFi connection is lost.   You don't want a WiFi connection issue to prevent you from turning off a light manually from a button attached to an ESP32.

Also, please feel free to add any new (or prior) PRs to the current dev branch for potential inclusion in next release.  The main focus of the next release is to integrate Ethernet support as an alternative to WiFi.

**Comment by frankonski** [comment]
> This ensures immediate and continuous manual operation of lights/switches when a device is rebooted or WiFi connection is lost. You don't want a WiFi connection issue to prevent you from turning off a light manually from a button attached to an ESP32.

hum.. I haven't thought of that. And I wonder if this is why my <brand I will not name> smart switches I paid $45 for are not responding to physical pushes... separate topic.

I will give the dev branch a try. Are you able to re-open PRs that were closed when I deleted my branch, like #880, and change the target branch to `dev`?

**Comment by frankonski** [comment]
I confirm the new method `setConnectionTimes()` is working fine for me now, no longer any errors logged.

I used `setConnectionTimes(10,60,5)`.

One small improvement to your documentation: just clarify this needs to be called prior to calling `homeSpan.begin()`;

**Comment by HomeSpan** [comment]
Glad to hear it's working.  But do you have problems if called before homeSpan.begin()?  That should not cause an issue since the WiFi process does not kick off until polling starts.

**Comment by frankonski** [comment]
>  But do you have problems if called before homeSpan.begin()?
No, that is what I suggested. Do I have problems if I call it after? No idea, didn't try. I assume it had to called before calling homeSpan.begin() even though it was not documented.

**Comment by HomeSpan** [comment]
Sorry, I meant "after" above.  The setConnectionTimes() should be able to be called either before or after homeSpan.begin(), as long as it is called before polling begins (which normally starts in the loop() function unless you start it in setup() with autopolling()).

In a future release I may combine begin() and autoPolling() into a single call that basically "starts" everything at the end of setup().  But just a nebulous idea for now....

**Comment by frankonski** [comment]
cool, thanks. Feel free to closed with your `Fixed in dev` tag :)

**Comment by HomeSpan** [comment]
Re-opening so I can reference in documentation for latest update.  Will close again once everything is released in next update.

**Comment by dexter12133** [comment]
I have the same issue on my ESP32-S3-devkitc (chinesse one). When ESP32 is connected only to USB - it is working fine, but after connecting it into Input/output board(own project), there is something like WIFI freezing. It cannot connect to WIFI and showing the same errors like in this topic. The interesting thing is that my other ESP32 S3 devboard is working fine on this IO Board... It looks like connecting any GPIO to board causing WIFI problems... Do you have similar problems before?

**Comment by HomeSpan** [comment]
If everything works fine when you *don't* connect to your IO board, but then it stops working when you do connect to your IO board, and you make absolutely no changes to the software when you connect to the IO board, then the problem is likely in your IO board either using pins that it should not, or perhaps drawing too much power.  Also, how do you know for sure the WiFi is freezing?  Have you tried pinging the board or using the Web Log log and verified indeed the WiFi is freezing and it is unrelated to HomeKit not responding?

The reason I ask is that if WiFi itself is working BUT you are getting no-response from HomeKit AND the issue only occurs then you have the IO board plugged in, do you have a sensor on the IO board that is feeding something back to HomeKit that is out-of-range?  If so, you will get a warning message in the Serial Monitor.

A lot can be going wrong, but you need to provide a lot more details about what you are seeing to help debug.

**Comment by dexter12133** [comment]
This is really strange, beacause I have two types of ESP32 devboards from aliexpress(both n16r8). One type of devboard is working fine, but the second type works like described. Today I'll check connecting of which pin causing these problem. Tomorrow I'll check power draw because it can be good clue. I'll also check Web log, because i've didn't use it yet.

**Comment by dexter12133** [comment]
I've just discovered that ESP is connecting to WIFI one I've touched wifi antena... So for this I've found that some people are lowering TX power, some people desoldered onboard LED... I guess that some capacitance appears on connecter PCB traces and that causing problems, but this is HW problem, not your library ;)

**Comment by HomeSpan** [comment]
Feature is now available in latest release v2.1.0.
